docs/spec/fastsync/fastsync.tla

```diff
@@ -0,0 +1,661 @@
+*)
+HandleStatusResponse(msg, bPool) ==
+    LET peerHeight == bPool.peerHeights[msg.peerId] IN
+
+    IF /\ msg.peerId \in bPool.peerIds
+       /\ msg.height >= peerHeight
+    THEN    \* a correct response
+        LET pHeights == [bPool.peerHeights EXCEPT ![msg.peerId] = msg.height] IN
+        [bPool EXCEPT !.peerHeights = pHeights]
+    ELSE RemovePeers({msg.peerId}, bPool)   \* the peer has sent us message with smaller height or peer is not in our peer list
+
+
+(*
+Handle BlockResponse message.
+If valid block response, update blockStore, pendingBlocks and receivedBlocks.
+If invalid (unsolicited response or malformed block), then remove peer.
+Returns new block pool.
+See https://github.com/tendermint/tendermint/blob/dac030d6daf4d3e066d84275911128856838af4e/blockchain/v2/scheduler.go#L522
+*)
+HandleBlockResponse(msg, bPool) ==
+    LET h == msg.block.height IN
+
+    IF /\ msg.peerId \in bPool.peerIds
+       /\ bPool.blockStore[h] = NilBlock
+       /\ bPool.pendingBlocks[h] = msg.peerId
+       /\ msg.block.wellFormed
+    THEN
+        [bPool EXCEPT
+            !.blockStore = [bPool.blockStore EXCEPT ![h] = msg.block],
+            !.receivedBlocks = [bPool.receivedBlocks EXCEPT![h] = msg.peerId],
+            !.pendingBlocks = [bPool.pendingBlocks EXCEPT![h] = NilPeer]
+         ]
+    ELSE RemovePeers({msg.peerId}, bPool)
+
+\* Compute max peer height.
+\* See https://github.com/tendermint/tendermint/blob/dac030d6daf4d3e066d84275911128856838af4e/blockchain/v2/scheduler.go#L440
+MaxPeerHeight(bPool) ==
+    IF bPool.peerIds = {}
+    THEN 0 \* no peers, just return 0
+    ELSE LET Hts == {bPool.peerHeights[p] : p \in bPool.peerIds} IN
+           CHOOSE max \in Hts: \A h \in Hts: h <= max
+
+(* Returns next height for which request should be sent.
+   Returns NilHeight in case there is no height for which request can be sent.
+   See https://github.com/tendermint/tendermint/blob/dac030d6daf4d3e066d84275911128856838af4e/blockchain/v2/scheduler.go#L454 *)
+FindNextRequestHeight(bPool) ==
+    LET S == {i \in Heights:
+                /\ i >= bPool.height
+                /\ i <= MaxPeerHeight(bPool)
+                /\ bPool.blockStore[i] = NilBlock
+                /\ bPool.pendingBlocks[i] = NilPeer} IN
+    IF S = {}
+        THEN NilHeight
+    ELSE
+        CHOOSE min \in S:  \A h \in S: h >= min
+
+\* Returns number of pending requests for a given peer.
+NumOfPendingRequests(bPool, peer) ==
+    LET peerPendingRequests ==
+        {h \in Heights:
+            /\ h >= bPool.height
+            /\ bPool.pendingBlocks[h] = peer
+        }
+    IN
+    Cardinality(peerPendingRequests)
+
+(* Returns peer that can serve block for a given height.
+   Returns NilPeer in case there are no such peer.
+   See https://github.com/tendermint/tendermint/blob/dac030d6daf4d3e066d84275911128856838af4e/blockchain/v2/scheduler.go#L477 *)
+FindPeerToServe(bPool, h) ==
```

Comment:
We make many calls to NumOfPendingRequests in here. Would it be worth deduplicating by calling it once for each peer and storing that in a local map variable and referencing it as needed?

How can we check what functions are worth trying to refactor? Goes back to our discussion about profiling from the Apalache call last week ...

Reply:
This is really a good point as inside NumOfPendingRequests we call Cardinality which according to Igor is not very model-checker friendly. @konnov

Reply:
I agree. If we can get rid of `Cardinality`, it would help Apalache, though it does not affect TLC that much. The question is whether we want to go one level of abstraction lower by introducing that additional map.